addon/file.ts

```diff
@@ -0,0 +1,323 @@
+   * range of 0 to 100.
+   */
+  @tracked progress = 0;
+
+  /**
+   * The current state that the file is in.
+   */
+  @tracked private internalState: FileState = FileState.Queued;
+
+  get state() {
+    return this.internalState;
+  }
+
+  set state(value) {
+    this.internalState = value;
+    this.queue?.flush();
+  }
+
+  // @TODO: For both `upload*()` methods, the return value `Promise<unknown>` is
+  // very much unspecified and needs more details
+
+  // /**
+  //   The source of the file. This is useful
+  //   for applications that want to gather
+  //   analytics about how users upload their
+  //   content.
+
+  //   This property can be one of the following:
+
+  //   - `browse`
+  //   - `drag-and-drop`
+  //   - `web`
+  //   - `data-url`
+  //   - `blob`
+
+  //   `browse` is the source when the file is created
+  //   using the native file picker.
+
+  //   `drag-and-drop` is the source when the file was
+  //   created using drag and drop from their desktop.
+
+  //   `web` is the source when the file was created
+  //   by dragging the file from another webpage.
+
+  //   `data-url` is the source when the file is created
+  //   from a data URL using the `fromDataURL` method for
+  //   files. This usually means that the file was created
+  //   manually by the developer on behalf of the user.
+
+  //   `blob` is the source when the file is created
+  //   from a blob using the `fromBlob` method for
+  //   files. This usually means that the file was created
+  //   manually by the developer.
+
+  //   @accessor source
+  //   @type {String}
+  //   @default ''
+  //   @readonly
+  //  */
+  // source?: FileSource;
+
+  /**
+   * Upload file with `application/octet-stream` content type.
+   *
+   * @param url Your server endpoint where to upload the file
+   * @param options additional request options
+   */
+  uploadBinary(
+    url: string,
+    options: Record<string, unknown>
+  ): Promise<unknown> {
+    options.contentType = 'application/octet-stream';
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore this is not typed
+    return upload(this, url, options, (request) => {
+      return request.send(this.#blob);
+    });
+  }
+
+  /**
+   * Upload file to your server
+   *
+   * @param url Your server endpoint where to upload the file
+   * @param options additional options, eg. `{ fileKey: string, data: { key: string } }`
+   */
+  upload(url: string, options: Record<string, unknown>): Promise<unknown> {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore this is not typed
+    return upload(this, url, options, (request, opts) => {
+      // Build the form
+      const form = new FormData();
+
+      for (const key of Object.keys(opts.data)) {
+        if (key === opts.fileKey) {
+          // @TODO: Am I doing this right here?
+          form.append(key, this.#blob, this.name);
+        } else {
+          form.append(key, opts.data[key]);
```

Comment:
Need help on the first `form.append()`. They were both thhe same code in current master but I was changing the one in the first if condition until tests were happy. Am I doing this right?

Reply:
Hopefully resolved above ☝🏻 . Didn't require a code change.